protocols/relay/src/v2/client/transport.rs

```diff
@@ -0,0 +1,411 @@
+    ///     PeerId::random(),
+    /// );
+    ///
+    /// // To reduce unnecessary connection attempts, put `relay_transport` first.
+    /// let transport = OrTransport::new(relay_transport, actual_transport);
+    /// ```
+    pub(crate) fn new() -> (Self, mpsc::Receiver<TransportToBehaviourMsg>) {
+        let (to_behaviour, from_transport) = mpsc::channel(0);
+
+        (ClientTransport { to_behaviour }, from_transport)
+    }
+}
+
+impl Transport for ClientTransport {
+    type Output = RelayedConnection;
+    type Error = RelayError;
+    type Listener = RelayListener;
+    type ListenerUpgrade = Ready<Result<Self::Output, Self::Error>>;
+    type Dial = RelayedDial;
+
+    fn listen_on(self, addr: Multiaddr) -> Result<Self::Listener, TransportError<Self::Error>> {
+        match parse_relayed_multiaddr(addr)? {
+            // Address does not contain circuit relay protocol.
+            Err(addr) => return Err(TransportError::MultiaddrNotSupported(addr)),
+            // Address does contain circuit relay protocol.
+            Ok(relayed_addr) => {
+                let (relay_peer_id, relay_addr) = match relayed_addr {
+                    RelayedMultiaddr {
+                        relay_peer_id: None,
+                        relay_addr: _,
+                        ..
+                    } => return Err(RelayError::MissingDstPeerId.into()),
+                    RelayedMultiaddr {
+                        relay_peer_id: _,
+                        relay_addr: None,
+                        ..
+                    } => return Err(RelayError::MissingRelayAddr.into()),
+                    RelayedMultiaddr {
+                        relay_peer_id: Some(peer_id),
+                        relay_addr: Some(addr),
+                        ..
+                    } => (peer_id, addr),
+                };
+
+                let (to_listener, from_behaviour) = mpsc::channel(0);
+                let mut to_behaviour = self.to_behaviour;
+                let msg_to_behaviour = Some(
+                    async move {
+                        to_behaviour
+                            .send(TransportToBehaviourMsg::ListenReq {
+                                relay_peer_id,
+                                relay_addr,
+                                to_listener,
+                            })
+                            .await
+                    }
+                    .boxed(),
+                );
+
+                Ok(RelayListener {
+                    queued_new_addresses: Default::default(),
+                    from_behaviour,
+                    msg_to_behaviour,
+                })
+            }
+        }
+    }
+
+    fn dial(self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        match parse_relayed_multiaddr(addr)? {
+            // Address does not contain circuit relay protocol.
+            Err(addr) => return Err(TransportError::MultiaddrNotSupported(addr)),
```

Comment:
This could be a `.map_err` followed by another `?`. At the point, my question would be why `parse_relayed_multiaddr` doesn't return a `TransportError` right away?

Reply:
Applied all suggestions via ed79c72ae67a4acec6f8bfd58d379a5bd0b956cf. Simplifications should have been included in 08d6637732d0e792e69f0f86e26c12ca5bc3b8e7 directly.